DOCSP-43069: Replace documents (#52)
mongodb/docs-c: source/write/replace.txt

.. _c-write-replace:

=================
Replace Documents
=================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: modify, change, code example

Overview
--------

> In this guide, you can learn how to use the {+driver-short+} to run a replace
> operation on a MongoDB collection. A replace operation performs differently

Reviewer: ```suggestion
In this guide, you can learn how to use the {+driver-short+} to run a **replace
operation** on a MongoDB collection. A replace operation performs differently
```

than an update operation. An update operation modifies only the specified
fields in a target document. A replace operation removes *all* fields
in the target document and replaces them with new ones.

To replace a document, use the ``mongoc_collection_replace_one()`` function.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

Replace Operation
-----------------

You can perform a replace operation by using ``mongoc_collection_replace_one()``. 
This function removes all fields except the ``_id`` field from the first document that
matches the search criteria. It then inserts the fields and values you specify into the 
document.

Required Parameters
~~~~~~~~~~~~~~~~~~~

The ``mongoc_collection_replace_one()`` function requires the following parameters:

> - Collection: Specifies the collection to update.

Reviewer: ```suggestion
- Collection: Specifies which collection to perform the replace operation on.
```

> - Query filter document: Specifies which collection documents to update. For
>   more information about query filters, see the 
>   :manual:`Query Filter Documents section </core/document/#query-filter-documents>` in
>   the {+mdb-server+} manual.

Reviewer: ```suggestion
- Query filter document: Specifies which collection documents to match. The function selects the first
  matching document to replace. For
  more information about query filters, see the 
  :manual:`Query Filter Documents section </core/document/#query-filter-documents>` in
  the {+mdb-server+} manual.
```

- Replacement document: Specifies the fields and values to insert into the new 
  document.

- Results location: Specifies a pointer to overwritable storage that will contain
  operation results, or ``NULL``.

- Error location: Specifies a location for an error value, or ``NULL``.

Example
~~~~~~~

> The following example uses the ``mongoc_collection_replace_one()`` function to replace the fields and values
> of a document in which the ``name`` field value is ``"Pizza Town"``:

Reviewer: ```suggestion
The following example uses the ``mongoc_collection_replace_one()`` function to replace the fields and values
of a document in which the ``name`` field value is ``"Pizza Town"`` with a document in which the name is ...:
```

.. literalinclude:: /includes/write/replace.c
    :language: c
    :dedent:
    :start-after: start-replace-one
    :end-before: end-replace-one

.. important::

   The values of ``_id`` fields are immutable. If your replacement document specifies 

>    a value for the ``_id`` field, it must match the ``_id`` value of the existing document.

Reviewer: ```suggestion
   a value for the ``_id`` field, it must be identical to the ``_id`` value of the existing document.
```

       | Defaults to ``false``.

   * - ``bypassDocumentValidation``
     - | Specifies whether the replace operation bypasses document validation. This lets you 
         replace documents that don't meet the schema validation requirements, if any 
         exist. For more information about schema validation, see :manual:`Schema
         Validation </core/schema-validation/#schema-validation>` in the MongoDB
         Server manual.
       | Defaults to ``false``.

   * - ``collation``
     - | Specifies the kind of language collation to use when comparing text.
         For more information, see :manual:`Collation </reference/collation/#std-label-collation>`
         in the {+mdb-server+} manual.

   * - ``hint``
     - | Gets or sets the index to scan for documents. 
         For more information, see the :manual:`hint statement </reference/command/update/#std-label-update-command-hint>`
         in the {+mdb-server+} manual.

   * - ``comment``
     - | Attaches a comment to the operation. For more information, see the :manual:`insert command
         fields </reference/command/insert/#command-fields>` guide in the
         {+mdb-server+} manual.

> Example
> ~~~~~~~

Reviewer: ```suggestion
Replace Options Example
~~~~~~~~~~~~~~~~~~~~~~~
```
